drivers/gpio/gpio_silabs.c

```diff
@@ -0,0 +1,416 @@
+
+	sl_hal_gpio_set_port(config->gpio_index, mask);
+
+	return 0;
+}
+
+static int gpio_silabs_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t mask)
+{
+	const struct gpio_silabs_port_config *config = dev->config;
+
+	sl_hal_gpio_clear_port(config->gpio_index, mask);
+
+	return 0;
+}
+
+static int gpio_silabs_port_toggle_bits(const struct device *dev, gpio_port_pins_t mask)
+{
+	const struct gpio_silabs_port_config *config = dev->config;
+
+	sl_hal_gpio_toggle_port(config->gpio_index, mask);
+
+	return 0;
+}
+
+static int gpio_silabs_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+					       enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	const struct gpio_silabs_port_config *config = dev->config;
+	struct gpio_silabs_port_data *data = dev->data;
+	struct gpio_silabs_common_data *common = data->common_dev->data;
+	sl_gpio_t gpio = {.port = config->gpio_index, .pin = pin};
+	int32_t int_no = SL_GPIO_INTERRUPT_UNAVAILABLE;
+	sl_gpio_interrupt_flag_t flag = SL_GPIO_INTERRUPT_RISING_FALLING_EDGE;
+
+	if (mode == GPIO_INT_MODE_LEVEL) {
+		LOG_ERR("Level interrupt not supported on pin %u", pin);
+		return -ENOTSUP;
+	}
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		ARRAY_FOR_EACH(common->irq_pin_map, i) {
+			if ((data->int_enabled_mask & BIT(i)) && (common->irq_pin_map[i] == pin)) {
+				sl_hal_gpio_disable_interrupts(BIT(i));
+				WRITE_BIT(data->int_enabled_mask, i, false);
+				common->irq_pin_map[i] = 0xFF;
+				break;
+			}
+		}
+	} else {
+		/* Check if pin already has an interrupt configured */
+		ARRAY_FOR_EACH(common->irq_pin_map, i) {
+			if ((data->int_enabled_mask & BIT(i)) && (common->irq_pin_map[i] == pin)) {
+				int_no = i;
+				break;
+			}
+		}
+
+		if (trig == GPIO_INT_TRIG_LOW) {
+			flag = SL_GPIO_INTERRUPT_FALLING_EDGE;
+		} else if (trig == GPIO_INT_TRIG_HIGH) {
+			flag = SL_GPIO_INTERRUPT_RISING_EDGE;
+		} else {
+			flag = SL_GPIO_INTERRUPT_RISING_FALLING_EDGE;
+		}
+
+		int_no = sl_hal_gpio_configure_external_interrupt(&gpio, int_no, flag);
+
+		if (int_no == SL_GPIO_INTERRUPT_UNAVAILABLE) {
+			LOG_ERR("No available interrupt for pin %u", pin);
+			return -EINVAL;
+		}
+
+		common->irq_pin_map[int_no] = pin;
+
+		sl_hal_gpio_enable_interrupts(BIT(int_no));
+		WRITE_BIT(data->int_enabled_mask, int_no, true);
+	}
+
+	return 0;
+}
+
+static int gpio_silabs_port_manage_callback(const struct device *dev, struct gpio_callback *cb,
+					    bool set)
+{
+	struct gpio_silabs_port_data *data = dev->data;
+
+	return gpio_manage_callback(&data->callbacks, cb, set);
+}
+
+static void gpio_silabs_common_isr(const struct device *dev)
+{
+	struct gpio_silabs_common_data *data = dev->data;
+	uint32_t int_status = GPIO->IF;
+
+	ARRAY_FOR_EACH(data->ports, i) {
+		if (!int_status) {
+			break;
+		}
+		const struct device *port_dev = data->ports[i];
+		struct gpio_silabs_port_data *port_data = port_dev->data;
+		struct gpio_silabs_common_data *common = port_data->common_dev->data;
```

Comment:
This can be hoisted outside the for loop -- `dev` is already the gpio peripheral (same as `port_data->common_dev`).